docs/async_ops.md

```diff
@@ -0,0 +1,106 @@
+# Async HLO Instructions
+
+1. Adding async operations to HLO is cumbersome (i.e. `all-reduce-start` and
+   `all-reduce-done`).
+2. The start and done split may be inadequate for some of the asynchronous use
+   cases.
+
+To target the first shortcoming, we propose to introduce one last set of new
+asynchronous opcodes: `kAsyncStart`, `kAsyncUpdate`, and `kAsyncDone``. The idea
+is to create a generic asynchronous opcode that can wrap any HLO instruction.
+The actual operation that will be performed asynchronously will be encoded using
+a called computation that only has the instruction as its root and any
+parameters for inputs. The in-flight input/output buffer handling and aliasing
+can then be shared for any asynchronous operation. The async-start instruction’s
+output shape will then be a tuple of the input operands, output values, and any
+intermediate state that is needed for the `async-update` or `async-done`
+instructions.
```

Comment:
@ezhulenev 
So in the end, we have one sync HLO for each NCCL call and `kAsyncStart`, `kAsyncUpdate`, and `kAsyncDone` to convert all of them to async mode.
Is that a good summary of the end goal?